### PR TITLE
Added support for transactions

### LINKED ISSUE #133: Add SQL transactions support
This should be implemented in `Sql::Query`.
Here's the related Sequel documentation: http://sequel.jeremyevans.net/rdoc/files/doc/transactions_rdoc.html

I'd say to not support/document `Sequel::Rollback`. Using exceptions for control flow is always a poor choice. If an exception is raised as side effect, the transaction should be rolled back.

```ruby
transaction do
  raise `Sequel::Rollback` # this will rollback anyway because of Sequel, but it's a poor style that we shouldn't encourage
end
```

```ruby
transaction do
  some_dangerous_method! # if this raises an exception, the transaction should be rolled back
end
```

The second style is the one that we should encourage/document.

/cc @stevehodgkiss @joneslee85

### PR DESCRIPTION
**Pull request by AlfonsoUceda**
Closes #133

### CODE REVIEW

**Comment by jodosha** [comment]
@AlfonsoUceda Thanks for this PR.

This looks good, but we need to test the type of transactions that we are going to support. Sequel has fairly advanced features for this. We're not helping enough developers to understand how to use those features and protecting them from changes.

### Help

I'm for expanding API docs with the options that we support and with examples.

### Changes

We depend on `sequel ~> 4.9`. If one of those options change in a minor version we cannot guarantee that Lotus::Model works with it.

```ruby
UserRepository.transaction(retry_on: [RecordNotUnique]) do
  UserRepository.create(User.new(username: 'jodosha'))
end
```

Now suppose that a new version of Sequel renames that `:retry_on` with `:retry`, the applications will suddenly break without us to even notice. The reason is we don't have tests to cover those cases.

The result is a breaking change for developers even if we haven't bumped to a version that allows us to do so. In other words: they depend on `lotus-model` which should guarantee that the application will work until they don't upgrade to a new version of it.

### Adapters interoperability

In order to increase the compatibility between adapters, I'm for removing that exception on the memory adapter and to document this limitation. The implementation should just yield the given block and do nothing more.



Can you please expand this PR? Thank you very much.

**Comment by AlfonsoUceda** [comment]
If I undestood well, we have to test every option that sequel 4.9 support in transactions to ensure breaks if sequel is updated.

In the `Adapters interoperability` section, I comment in code.

File: lib/lotus/model/adapters/abstract.rb [hunk 1]
```diff
@@ -162,6 +170,15 @@ def command(query)
...
           raise NotImplementedError
         end
+
+        # Create a transaction
+        #
+        # @param options [Hash] options for transaction
+        #
+        # @since x.x.x
+        def transaction(options = {})
+          raise MethodNotSupported
```
Comment (AlfonsoUceda) [inline lib/lotus/model/adapters/abstract.rb]: remove this Exception and only yield? and document that only available in sql adapter?

File: test/repository_test.rb [hunk 2]
```diff
@@ -299,4 +299,61 @@ def to_int
...
+        ArticleRepository.all.size.must_equal 1
+        ArticleRepository.transaction do
+          ArticleRepository.create(article2)
+        end
+        ArticleRepository.all.size.must_equal 2
+      end
+    end
+  end
+
+  describe "with memory adapter" do
```
Comment (AlfonsoUceda) [inline test/repository_test.rb]: don't test this case, isn't it?

**Comment by jodosha** [comment]
@AlfonsoUceda Yes. My point is: we should test the options that we want to support. I'd say we can start with the following list for the next release:

* `rollback: :reraise`
* `rollback: :always`

So the API docs and the tests should be explicit on those cases. That's it :)

File: test/repository_test.rb (hunk 2 above)
Comment (jodosha) [inline test/repository_test.rb]: I'd say to keep the test to document the existence of `MemoryAdapter#transaction`, but the test should not to reproduce SQL behavior (rollback), only assert what it does.

```ruby
-> {
BookRepository.transaction do
  raise "boom"
end
}.must_raise RuntimeError
```

;)

File: lib/lotus/model/adapters/abstract.rb (hunk 1 above)
Comment (jodosha) [inline lib/lotus/model/adapters/abstract.rb]: :+1:

**Comment by AlfonsoUceda** [comment]
@jodosha what do you think about these changes?

**Comment by jodosha** [comment]
@AlfonsoUceda This looks good, I'll do a minor API docs cleanup. Thank you! :+1:

**Comment by AlfonsoUceda** [comment]
You welcome :D

**Comment by jodosha** [comment]
@joneslee85 Can you please review this when you have time? TY!

File: test/repository_test.rb (hunk 2 above)
Comment (joneslee85) [inline test/repository_test.rb]: From user perspective, I think it's worth to raise something a meaningful message that the adapter does not support transaction rather than a generic RuntimeError. But that could be another PR.

**Comment by joneslee85** [comment]
Me happy :+1:

**Comment by jodosha** [comment]
@AlfonsoUceda Thank you, I've just merged this. :+1:
